### PR TITLE
c18n: Create ld-elf64cb-c18n.so.1 for the benchmark ABI

### PR DESCRIPTION
**Pull request by dpgao**
This PR adds the benchmark ABI-variant of the compartmentalising dynamic linker.

### CODE REVIEW

**Comment by jrtc27** [comment]
“c18n: Fix compartment tracing bug” is not an insightful commit message. What bug?

File: libexec/rtld-elf/aarch64/rtld_c18n_asm.S [hunk 1]
```diff
@@ -81,14 +81,13 @@ END(_rtld_sighandler)
...
-	 * This function MUST preserve all caller-saved registers except:
-	 * - c11: Target restricted stack
-	 * - c18: Will be overwritten later in the trampoline
+	 * This function MUST preserve all caller-saved registers.
 	 */
 	stp	c18, c30, [csp, #-(CAP_WIDTH * 2)]!
-	stp	c15, c16, [csp, #-(CAP_WIDTH * 2)]!
-	stp	c13, c14, [csp, #-(CAP_WIDTH * 2)]!
-	stp	c10, c12, [csp, #-(CAP_WIDTH * 2)]!
+	stp	c16, c17, [csp, #-(CAP_WIDTH * 2)]!
```
Comment (jrtc27) [inline libexec/rtld-elf/aarch64/rtld_c18n_asm.S:117]: Saving but not restoring c17 is weird

File: libexec/rtld-elf/aarch64/rtld_c18n_asm.S [hunk 2]
```diff
@@ -101,43 +100,64 @@ ENTRY(_rtld_get_rstk)
 
 	mov	w0, w14					/* w0 = cid */
 	bl	get_rstk
-	mov	c11, c0
+	mov	c17, c0
```
Comment (jrtc27) [inline libexec/rtld-elf/aarch64/rtld_c18n_asm.S:133]: I'm not a huge fan of all the non-standard calling conventions for the various hooks, in particular when there's no documentation about what's going on. Documentation for the standard calling convention isn't needed, because one expects that anything put in [cwx]0 is the return value, but using a random register like c17 to communicate between pieces of code *really* needs some documentation. Including for why c17 was picked (presumably because it's IP1 and can be used as a temporary register between function calls, though why not c16 i.e. IP0?...).

File: libexec/rtld-elf/aarch64/rtld_c18n_asm.S [hunk 3]
```diff
@@ -172,35 +192,8 @@ tramp_header_hook_def:
 	.chericap	0
 tramp_header_hook_function:
 	.chericap	0
```
Comment (jrtc27) [inline libexec/rtld-elf/aarch64/rtld_c18n_asm.S:228]: At this point, if neither of them has any code, just use a C struct for these things. The assembly doesn't reference the names, only the C does, so you gain nothing by having it in assembly.

File: libexec/rtld-elf/rtld_c18n.h [hunk 4]
```diff
@@ -122,6 +122,15 @@ untrusted_stk_set(void *sp)
 	asm ("msr	rcsp_el0, %0" :: "C" (sp));
 }
 
+static inline void *
+untrusted_stk_get(void)
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.h:142]: stk_table_get, stk_table_set, untrusted_stk_set and then untrusted_stk_get is an unusual order, though I guess the benchmark ABI commit makes it less weird because only one becomes trusted

File: libexec/rtld-elf/rtld_c18n.c [hunk 5]
```diff
@@ -142,7 +142,7 @@ static struct compart rtld_compart = {
 };
 
 static struct compart libthr_compart = {
-	.name = "libthr.so.3",
+	.name = "[TCB]",
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.c:145]: I mean it's still called libthr_compart... and is this really needed as part of the bug fix?

File: libexec/rtld-elf/rtld_c18n.c [hunk 6]
```diff
@@ -245,7 +245,7 @@ tramp_should_include(const Obj_Entry *reqobj, const struct tramp_data *data)
  * Set a dummy Restricted stack so that trampolines do not need to test if the
  * Restricted stack is valid.
  */
-static void *dummy_stack = &dummy_stack;
+static void *dummy_stack[1] = { &dummy_stack[0] };
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.c:249]: Why? T and T[1] are both a single object.

File: libexec/rtld-elf/rtld_c18n.c [hunk 7]
```diff
@@ -396,7 +396,7 @@ get_rstk(unsigned index)
 		table = stk_table_expand(table, capacity, true);
 		stk_table_set(table);
 	}
-	assert(table->stacks[cid_off].bottom == NULL);
+	assert(table->stacks[cid_off].size == 0);
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.c:401]: Is this related to the tracing bug?

File: libexec/rtld-elf/rtld_c18n.c [hunk 8]
```diff
@@ -1038,9 +1038,9 @@ _rtld_thr_exit_impl(long *state)
 	struct stk_table *table = stk_table_get();
 
 	for (size_t i = 0; i < table->capacity; ++i) {
-		stk = table->stacks[i].bottom;
-		if (stk != NULL) {
-			size = table->stacks[i].size;
+		size = table->stacks[i].size;
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.c:1059]: Is this related to the tracing bug?

File: libexec/rtld-elf/rtld_c18n.h [hunk 9]
```diff
@@ -127,7 +143,11 @@ untrusted_stk_get(void)
 {
 	void *sp;
 
+#ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
+	asm ("mov	%0, csp" : "=C" (sp));
```
Comment (jrtc27) [inline libexec/rtld-elf/rtld_c18n.h:147]: It looks like the only use of this doesn't care about the current address, but you need to be really careful when reading CSP because the compiler makes no guarantees about what value it will have (and if the asm isn't volatile then the assembly could be duplicated and see different addresses). I believe the current use doesn't matter, but I get nervous when C accesses CSP directly, rather than just having assembly pass it through to the C function like it used to.

File: libexec/rtld-elf/aarch64/rtld_c18n_asm.S (hunk 2 above)
Comment (dpgao) [reply libexec/rtld-elf/aarch64/rtld_c18n_asm.S:133]: c17 is used merely because it is the register allocated to hold the target restricted stack in the trampoline. It could have been any other temporary register. I will document it at the top of `_rtld_get_rstk`, which is only supposed to be called from the trampoline.

File: libexec/rtld-elf/rtld_c18n.c (hunk 5 above)
Comment (dpgao) [reply libexec/rtld-elf/rtld_c18n.c:145]: Good catch. This actually fixes another tracing bug where transitions to libc are being printed as transitions to libthr. I'll mention it in the commit message.

File: libexec/rtld-elf/rtld_c18n.c (hunk 6 above)
Comment (dpgao) [reply libexec/rtld-elf/rtld_c18n.c:249]: Just to make the intent clearer that this is a `stack` whose bottom element points to itself.

File: libexec/rtld-elf/rtld_c18n.h (hunk 9 above)
Comment (dpgao) [reply libexec/rtld-elf/rtld_c18n.h:147]: How about `cheri_getstack()`?

File: bin/cheribsdtest/Makefile [hunk 10]
```diff
@@ -23,6 +23,10 @@ SUBDIR+=	purecap-benchmark \
 		purecap-benchmark-dynamic \
 		purecap-benchmark-dynamic-mt \
 		purecap-benchmark-mt
+
+.if ${MACHINE_CPUARCH} == "aarch64" && ${MACHINE_ABI:Mpurecap}
+SUBDIR+=	benchmark-mt-c18n
```
Comment (jrtc27) [inline bin/cheribsdtest/Makefile:28]: Please also teach share/examples/cheribsdtest/run-many about this (installed to /usr/share/examples/cheribsdtest/run-many as a helpful script to run one or more cheribsdtest variants and filter out all lines other than the summary):

```diff
diff --git a/share/examples/cheribsdtest/run-many b/share/examples/cheribsdtest/run-many
index ffdcdf6f3da8..90cbd35e9fbf 100755
--- a/share/examples/cheribsdtest/run-many
+++ b/share/examples/cheribsdtest/run-many
@@ -45,7 +45,7 @@ if [ $# -eq 0 ]; then
 
 	case `uname -p` in
 	aarch64c)
-		set -- "$@" c18n
+		set -- "$@" c18n benchmark-c18n
 		;;
 	esac
 fi
@@ -58,8 +58,8 @@ fi
 suffixes=
 for abi in "$@"; do
 	case "$abi" in
-	c18n)
-		suffixes="$suffixes mt-c18n"
+	c18n|benchmark-c18n)
+		suffixes="$suffixes ${abi%c18n}mt-c18n"
 		;;
 	*)
 		for dyn in '' -dynamic; do
```
should work, I believe

File: libexec/rtld-elf/rtld_c18n.h (hunk 9 above)
Comment (jrtc27) [reply libexec/rtld-elf/rtld_c18n.h:147]: Same issue, it's not about the assembly, it's about CSP's address being ill-defined within a function. The code was fine and worked, I can just see issues in future if you start to use untrusted_stk_get in other ways, it was just an observation about that. Certainly doing `mov %0, csp` followed by `sp = cheri_getstack()` as you now do doesn't make sense, that'll just always overwrite the output of the assembly and make it dead...

File: libexec/rtld-elf/aarch64/rtld_c18n_asm.S [hunk 11]
```diff
@@ -44,51 +49,76 @@ ENTRY(_rtld_longjmp)
...
+#endif
 	b	_rtld_sighandler_impl
 END(_rtld_sighandler)
 
 ENTRY(_rtld_get_rstk)
 	/*
 	 * This function MUST preserve all caller-saved registers except:
-	 * - c11: Target restricted stack
-	 * - c18: Will be overwritten later in the trampoline
+	 * - c17: Target restricted stack
```
Comment (jrtc27) [inline libexec/rtld-elf/aarch64/rtld_c18n_asm.S:115]: This isn't hugely useful, it's the what not the why. If you want to document all the non-standard calling conventions going on, you need to say what registers are being used in non-standard ways on entry and exit, and where they're coming from / being passed onto. Otherwise anyone reading this code has no idea how it fits together and has to trawl through the assembly to figure out the flow of these random registers.

File: libexec/rtld-elf/rtld_c18n.h (hunk 9 above)
Comment (dpgao) [reply libexec/rtld-elf/rtld_c18n.h:147]: Oops I somehow forgot to delete the asm line. But I agree that this could be dangerous in the future so I'll revert to the old way to passing csp through assembly.

**Review by jrtc27 — APPROVED** [review]
No more objections from me, though again I haven't trawled through the assembly to check it in great detail. I assume you've verified that the updated run-many script does indeed run both cheribsdtest-mt-c18n and cheribsdtest-benchmark-mt-c18n correctly?
